test/contrib/test_socks.py

```diff
@@ -0,0 +1,492 @@
+
+    # Check some basic stuff.
+    assert received_version == SOCKS_VERSION_SOCKS4
+    assert command == b'\x01'  # Only support connect, not bind.
+
+    if username is not None and username != provided_username:
+        sock.sendall(b'\x00\x5d\x00\x00\x00\x00\x00\x00')
+        sock.close()
+        yield False
+        return
+
+    # Yield the address port tuple.
+    succeed = yield addr, port
+
+    if succeed:
+        response = b'\x00\x5a\xea\x60\x7f\x00\x00\x01'
+    else:
+        response = b'\x00\x5b\x00\x00\x00\x00\x00\x00'
+
+    sock.sendall(response)
+    yield True  # Avoid StopIteration exceptions getting fired.
+
+
+class TestSocks5Proxy(IPV4SocketDummyServerTestCase):
+    """
+    Test the SOCKS proxy in SOCKS5 mode.
+    """
+    def test_basic_request(self):
+        def request_handler(listener):
+            sock = listener.accept()[0]
+
+            handler = handle_socks5_negotiation(sock, negotiate=False)
+            addr, port = next(handler)
+
+            self.assertEqual(addr, '16.17.18.19')
+            self.assertTrue(port, 80)
+            handler.send(True)
+
+            while True:
+                buf = sock.recv(65535)
+                if buf.endswith(b'\r\n\r\n'):
+                    break
+
+            sock.sendall(b'HTTP/1.1 200 OK\r\n'
+                         b'Server: SocksTestServer\r\n'
+                         b'Content-Length: 0\r\n'
+                         b'\r\n')
+            sock.close()
+
+        self._start_server(request_handler)
+        proxy_url = "socks5://%s:%s" % (self.host, self.port)
+        pm = socks_proxy.SOCKSProxyManager(proxy_url)
+        response = pm.request('GET', 'http://16.17.18.19')
+
+        self.assertEqual(response.status, 200)
+        self.assertEqual(response.data, b'')
+        self.assertEqual(response.headers['Server'], 'SocksTestServer')
+
+    def test_correct_header_line(self):
+        def request_handler(listener):
+            sock = listener.accept()[0]
+
+            handler = handle_socks5_negotiation(sock, negotiate=False)
+            addr, port = next(handler)
+
+            self.assertEqual(addr, b'http2bin.org')
+            self.assertTrue(port, 80)
+            handler.send(True)
+
+            buf = b''
+            while True:
+                buf += sock.recv(65535)
+                if buf.endswith(b'\r\n\r\n'):
+                    break
+
+            self.assertTrue(buf.startswith(b'GET / HTTP/1.1'))
+            self.assertTrue(b'Host: http2bin.org' in buf)
+
+            sock.sendall(b'HTTP/1.1 200 OK\r\n'
+                         b'Server: SocksTestServer\r\n'
+                         b'Content-Length: 0\r\n'
+                         b'\r\n')
+            sock.close()
+
+        self._start_server(request_handler)
+        proxy_url = "socks5://%s:%s" % (self.host, self.port)
+        pm = socks_proxy.SOCKSProxyManager(proxy_url)
+        response = pm.request('GET', 'http://http2bin.org')
```

Comment:
Ah fair enough, in that case let's make it a non-URL (example.com or something).